src/graphics/canvas-layer.js

```diff
@@ -191,102 +202,118 @@ Crafty._registerLayerTemplate("Canvas", {
-            if (!overlap(rect, view)) continue;
+            if (!rect || !overlap(rect, view)) continue;
 
-            //search for ents under dirty rect
-            q = Crafty.map.search(rect, false);
+            // Draw the rectangle, collision search needs to check whether found entities overlap rect
+            this._drawRect(rect, true);
+        }
 
-            //clear the rect from the main canvas
-            ctx.clearRect(rect._x, rect._y, rect._w, rect._h);
+        // Draw dirty rectangles for debugging, if that flag is set
+        if (this.debugDirty === true) {
+            ctx.strokeStyle = 'red';
+            for (i = 0, l = dirty.length; i < l; ++i) {
+                rect = dirty[i];
+                ctx.strokeRect(rect._x, rect._y, rect._w, rect._h);
+            }
+        }
 
-            //Then clip drawing region to dirty rectangle
-            ctx.save();
-            ctx.beginPath();
-            ctx.rect(rect._x, rect._y, rect._w, rect._h);
-            ctx.clip();
+    },
 
-            // Loop over found objects removing dupes and adding visible canvas objects to array
-            for (j = 0, len = q.length; j < len; ++j) {
-                obj = q[j];
+    /**@
+     * #._drawDirtyCells
+     * @comp CanvasLayer
+     * @kind Method
+     * @private
+     *
+     * @sign public ._drawDirtyCells([Object screenRect])
+     * @param screenRect - a rectangular region {_x: x_val, _y: y_val, _w: w_val, _h: h_val}
+     *
+     * - If screenRect is omitted, redraw within the viewport
+     * - If screenRect is provided, redraw within the screenRect. It must contain final screen coordinates!
+     *
+     * - Triggered by the "RenderScene" event
+     * - This method is invoked if the number of rects is under 60% of the total number of objects
+     *  and the total number of objects is greater than 16.
+     * - Clear the dirty spatial grid cells, and redraw entities overlapping the dirty spatial grid cells.
+     *
+     * @see Canvas#.draw
+     */
+    _drawDirtyCells: function (view) {
+        view = view || this._viewportRect();
+        var i, l,
+            rect = this._tempRect,
+            dirtyRects = this._dirtyRects,
+            rectManager = Crafty.rectManager,
+            overlap = rectManager.overlap,
+            ctx = this.context;
 
-                if (dupes[obj[0]] || !obj._visible || (obj._drawLayer !== this) )
-                    continue;
-                dupes[obj[0]] = true;
-                objs.push(obj);
-            }
+        // Canvas works better with integral coordinates where possible
+        view = rectManager.integerBounds(view);
 
-            // Sort objects by z level
-            objs.sort(this._sort);
+        // Calculate dirty spatial map cells from all changed objects
+        this._createDirtyCells();
+        // Afterwards, calculate dirty rectangles from dirty spatial map cells
+        this._createDirtyRects();
```

Comment:
Ah yes, it could be changed so that dirty rectangles get created alongside dirty cell in a single iteration, if the current cell hasn't been marked as dirty yet.